util/rpcclient/rpcclient.go

```diff
@@ -0,0 +1,208 @@
+	if c.client == nil {
+		return errors.New("not connected")
+	}
+	logId := atomic.AddUint64(&c.logId, 1)
+	log.Trace("sending RPC request", "method", method, "logId", logId, "args", logArgs(int(c.config().ArgLogLimit), args...))
+	var err error
+	for i := 0; i < int(c.config().Retries)+1; i++ {
+		if ctx_in.Err() != nil {
+			return ctx_in.Err()
+		}
+		var ctx context.Context
+		var cancelCtx context.CancelFunc
+		timeout := c.config().Timeout
+		if timeout > 0 {
+			ctx, cancelCtx = context.WithTimeout(ctx_in, timeout)
+		} else {
+			ctx, cancelCtx = context.WithCancel(ctx_in)
+		}
+		err = c.client.CallContext(ctx, result, method, args...)
+		cancelCtx()
+		logger := log.Trace
+		limit := int(c.config().ArgLogLimit)
+		if err != nil && err.Error() != "already known" {
+			logger = log.Info
+			limit = 0
+		}
+		logger("rpc response", "method", method, "logId", logId, "result", limitString(limit, fmt.Sprintf("%+v", result)), "attempt", i, "args", logArgs(limit, args...))
+		if err == nil {
+			return nil
+		}
+		if errors.Is(err, context.DeadlineExceeded) {
+			continue
+		}
+		retryErrors := c.config().RetryErrors
+		if retryErrors != "" {
+			match, regexErr := regexp.MatchString(retryErrors, err.Error())
+			if regexErr != nil {
+				log.Warn("rpcclient: bad value for retry-error. Not retrying.", "err", err, "value", retryErrors)
+			}
+			if match {
+				continue
+			}
+		}
+		return err
+	}
+	return err
+}
+
+func (c *RpcClient) BatchCallContext(ctx context.Context, b []rpc.BatchElem) error {
+	return c.client.BatchCallContext(ctx, b)
+}
+
+func (c *RpcClient) EthSubscribe(ctx context.Context, channel interface{}, args ...interface{}) (*rpc.ClientSubscription, error) {
+	return c.client.EthSubscribe(ctx, channel, args...)
+}
+
+func (c *RpcClient) Start(ctx_in context.Context) error {
+	url := c.config().URL
+	jwtPath := c.config().JWTSecret
+	if url == "self" {
+		if c.autoStack == nil {
+			return errors.New("self not supported for this connection")
+		}
+		url = c.autoStack.WSEndpoint()
+		jwtPath = ""
+	} else if url == "self-auth" {
+		if c.autoStack == nil {
+			return errors.New("self-auth not supported for this connection")
+		}
+		url = c.autoStack.WSAuthEndpoint()
+		jwtPath = c.autoStack.JWTPath()
+	} else if url == "" {
+		return errors.New("no url provided for this connection")
+	}
+	var jwt *common.Hash
+	if jwtPath != "" {
+		var err error
+		jwt, err = signature.LoadSigningKey(jwtPath)
+		if err != nil {
+			return err
+		}
+	}
+	connTimeout := time.After(c.config().ConnectionWait)
+	for {
+		var ctx context.Context
+		var cancelCtx context.CancelFunc
+		timeout := c.config().Timeout
+		if timeout > 0 {
+			ctx, cancelCtx = context.WithTimeout(ctx_in, timeout)
+		} else {
+			ctx, cancelCtx = context.WithCancel(ctx_in)
+		}
+		var err error
+		var client *rpc.Client
+		if jwt == nil {
+			client, err = rpc.DialWebsocket(ctx, url, "")
```

Comment:
I think this should be `DialContext` not `DialWebsocket` -- as-is, it fails when an http URL is specified:
```
CRIT [05-22|14:07:44.324] couldn't connect to L1                   err="timeout trying to connect lastError: malformed ws or wss URL"
```